feat!: updated to use algokit-utils-py 5.0 types.

daniel-makerx opened the pull request:
Converted TypedDicts into dataclasses for clearer code and better intellisense.
General rewrite to be more pythonic

daniel-makerx committed:
chore: add missing py.typed file

daniel-makerx committed:
build: change dependency from sdk to algokit-utils-py 5.0 alpha

daniel-makerx committed:
feat!: updated to use algokit-utils-py 5.0 types.

Converted TypedDicts into dataclasses for clearer code and better intellisense

neilcampbell committed:
chore: update to the latest utils-py version

PatrickDinh commented on src/algokit_subscriber/types/subscription.py:
```diff
@@ -85,104 +108,125 @@ class BlockMetadata(TypedDict):
     """
 
-    transactions_root_sha256: str
+    transactions_root_sha256: bytes | None
```
these fields are base64 string in TS version

daniel-makerx replied:
These were base64 originally too, but the new typed response represent them as bytes so it seemed better to make them bytes here too.

I also just realized the `BlockMetadata` class could actually be removed and we could just use the algod `BlockHeader` directly, that would remove the mapping code and any future fields would come through automatically.

PatrickDinh commented on src/algokit_subscriber/types/arc28.py:
```diff
@@ -68,23 +99,23 @@ class Arc28EventToProcess(TypedDict):
-
-class EmittedArc28Event(Arc28EventToProcess):
+@dataclass(kw_only=True, slots=True)
+class EmittedArc28Event:
```
this type mismatches with the TS version
```
/** An emitted ARC-28 event extracted from an app call log. */
export interface EmittedArc28Event extends Arc28EventToProcess {
  /** The ordered arguments extracted from the event that was emitted */
  args: ABIValue[]
  /** The named arguments extracted from the event that was emitted (where the arguments had a name defined) */
  argsByName: Record<string, ABIValue>
}

/** An ARC-28 event to be processed */
export interface Arc28EventToProcess {
  /** The name of the ARC-28 event group the event belongs to */
  groupName: string
  /** The name of the ARC-28 event that was triggered */
  eventName: string
  /** The signature of the event e.g. `EventName(type1,type2)` */
  eventSignature: string
  /** The 4-byte hex prefix for the event */
  eventPrefix: string
  /** The ARC-28 definition of the event */
  eventDefinition: Arc28Event
}
```

daniel-makerx replied:
The `Arc28EventToProcess` class was only used internally previously so I just combined it with `EmittedArc28Event` to keep the public API smaller. Additionally having the event object on this seemed better than copying all the properties over. But happy to just change it to what TS had if thats simplest

PatrickDinh replied:
I'm incline making this the same with the TS version because:
- this class is exported to user. 👈 we can avoid a breaking change
- the TS version matches the Python version on `main` branch

PatrickDinh commented on src/algokit_subscriber/types/arc28.py:
```diff
@@ -1,65 +1,96 @@
+
+
+@dataclass(kw_only=True, slots=True)
+class Arc28EventGroup:
```
the field `group_name` is missing here compared to the TS version

daniel-makerx replied:
The group name became the key in the mapping in arc28_events

PatrickDinh replied:
I think we should add `group_name` back to avoid a breaking change. TS does the same.

PatrickDinh commented on src/algokit_subscriber/types/subscription.py:
```diff
@@ -192,176 +236,236 @@ class BeforePollMetadata(TypedDict):
 
-    arc28_events: NotRequired[list["Arc28EventGroup"]]
-    """Any ARC-28 event definitions to process from app call logs"""
+    arc28_events: Mapping[str, Arc28EventGroup] | None = None
```
on `main` and in the TS version, this is a list

daniel-makerx replied:
Other parts of the API assume the event group name is unique, so it seemed a better fit to make this a mapping. But if we want to keep TS and PY closely aligned then we can change this back to a list

PatrickDinh replied:
I'm leaning toward that we shouldn't introduce a breaking change here too.

PatrickDinh commented on src/algokit_subscriber/types/subscription.py:
```diff
@@ -16,45 +16,68 @@ class BalanceChangeRole(Enum):
+
+
+@dataclass(kw_only=True, slots=True)
+class BlockMetadata:
```
compared to the TS version, these fields are missing
```
  /** Tracks the status of state proofs. */
  stateProofTracking?: BlockStateProofTracking[]
  /** Fields relating to voting for a protocol upgrade. */
  upgradeVote?: BlockUpgradeVote
  /** Participation account data that needs to be checked/acted on by the network. */
  participationUpdates?: ParticipationUpdates
  /** Address of the proposer of this block */
  proposer?: string
```

daniel-makerx replied:
As mentioned before I think it might make sense to just replace this type with the algod `BlockHeader`

PatrickDinh replied:
@daniel-makerx I had a look, it isn't straight forward to replace it with `BlockHeader` due to some computed fields:
- hash - extracted from certificate (not in BlockHeader)
- parent_transaction_count - computed from payset length
- full_transaction_count - computed recursively including inner txns
- transactions_root / transactions_root_sha256 - extracted from txn_commitments

Updating these fields will be a breaking change

PatrickDinh committed:
refactor: align with TS

PatrickDinh committed:
refactor: align with main

PatrickDinh committed:
fix: include asset_config and asset_freeze in asset_id filter

The _get_txn_asset_id function was missing checks for asset_config_transaction
and asset_freeze_transaction, causing asset_id filters to incorrectly exclude
asset modify/destroy (acfg) and asset freeze (afrz) transactions.

PatrickDinh committed:
refactor: remove redundant assertion on id_ field

The id_ field in SubscribedTransaction is now a required str (not str | None),
making the assertion unnecessary.

PatrickDinh committed:
refactor: remove _arc28_events_to_dict helper function

PatrickDinh committed:
fix: events filter bug

PatrickDinh commented:
hey @daniel-makerx, thanks for the responses, they are very helpful and giving me context. I'll update the PR #83 to revert some of my changes and use `BlockHeader` directly.

PatrickDinh committed:
refactor: add NamedTransactionFilter back

PatrickDinh committed:
chore: docs

PatrickDinh committed:
chore: update README

PatrickDinh committed:
PR review (#83)

## Proposed Changes

- added `NamedTransactionFilter` class back to avoid breaking change
- Refactored `arc28_events` from `Mapping[str, Arc28EventGroup]` to
`list[Arc28EventGroup]` with `group_name` as a field on the dataclass,
aligning with the TypeScript implementation
- Flattened `EmittedArc28Event` to expose `group_name`, `event_name`,
`event_signature`, and `event_prefix` directly instead of nested `group`
and `event` objects
- Added missing `BlockMetadata` fields: `proposer`,
`state_proof_tracking`, `upgrade_vote`, `participation_updates`
- Changed `transactions_root` and `transactions_root_sha256` from
`bytes` to base64-encoded `str`
- Fixed `asset_id` filter to include `asset_config` and `asset_freeze`
transactions
- Removed redundant assertion on `id_` field in deduplication

PatrickDinh committed:
chore: update utils alpha (#84)

PatrickDinh committed:
feat: support reject version

PatrickDinh committed:
feat: support reject version (#85)

## Proposed Changes

-
-
-
